### PR TITLE
Add Ballast To Weather App

### PR DESCRIPTION
**Pull request by RamziJabali**
## Android:

https://user-images.githubusercontent.com/18749441/179202137-52492f1a-1795-4264-a3ad-835addd2b172.mp4

## iOS:

https://user-images.githubusercontent.com/18749441/179201090-476078a9-2451-4c3b-b52e-05c43ad875e2.mov

### 👷‍♂️ Testing
- `./gradlew :app:build`
- `./gradlew :shared:build`
- `xcodebuild -workspace ios/KaMPKitiOS.xcworkspace -scheme KaMPKitiOS 
    -sdk iphoneos -configuration Debug build -destination name="iPhone 8"`

### CODE REVIEW

**Review by seljabali — APPROVED** [review]

File: gradle/libs.versions.toml [hunk 1]
```diff
@@ -8,17 +8,23 @@ compileSdk = "31"
...
 moko = "0.12.0"
+ballast = "1.2.1"
 
 [libraries]
 
+ballast-core = {module = "io.github.copper-leaf:ballast-core", version.ref = "ballast"}
+ballast-saved-state = {module = "io.github.copper-leaf:ballast-saved-state", version.ref = "ballast"}
+ballast-repository = {module = "io.github.copper-leaf:ballast-repository", version.ref = "ballast"}
+ballast-debugger = {module = "io.github.copper-leaf:ballast-debugger", version.ref = "ballast"}
+
```
Comment (seljabali) [inline gradle/libs.versions.toml:27]: Remove below:
```
orbit-mvi-core = { module = "org.orbit-mvi:orbit-core", version.ref = "orbit" }
orbit-mvi-viewmodel = { module = "org.orbit-mvi:orbit-viewmodel", version.ref = "orbit" }
```

Also do we still need icerock?
```
icerock-moko-mvvm-core = {module = "dev.icerock.moko:mvvm-core", version.ref = "moko"}
```

File: ios/KaMPKitiOS/BreedListScreen.swift [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+//  Copyright © 2021 Touchlab. All rights reserved.
+//
+
+import Combine
+import SwiftUI
+import shared
+
+private let log = koin.loggerWithTag(tag: "ViewController")
+
+struct BreedListScreen: View {
```
Comment (seljabali) [inline ios/KaMPKitiOS/BreedListScreen.swift:15]: Remove

File: ios/KaMPKitiOS/Resources/Colors.swift [hunk 3]
```diff
@@ -1,18 +1,18 @@
...
-    static let darkBlue = UIColor(hex: 0x00008B)!
-    static let gainsboro = UIColor(hex: 0xDCDCDC)!
-    static let yellow = UIColor(hex: 0xFFFF00)!
-    static let purple200 = UIColor(hex: 0xFFBB86FC)!
-    static let purple500 = UIColor(hex: 0xFF6200EE)!
-    static let purple700 = UIColor(hex: 0xFF3700B3)!
-    static let teal200 = UIColor(hex: 0xFF03DAC5)!
-    static let dimGray = UIColor(hex: 0xFF696969)!
-}
+//import Foundation
```
Comment (seljabali) [inline ios/KaMPKitiOS/Resources/Colors.swift:1]: Why are we commenting this out?

File: ios/Podfile [hunk 4]
```diff
@@ -6,5 +6,5 @@ install! 'cocoapods', :deterministic_uuids => false
 
 target 'KaMPKitiOS' do
     pod 'shared', :path => '../shared/'
-    pod 'SwifterSwift/SwiftStdlib', '~> 5.2.0' # Standard Library Extensions
+    pod 'SwiftLint'
```
Comment (seljabali) [inline ios/Podfile:9]: Add version

File: shared/build.gradle.kts [hunk 5]
```diff
@@ -59,13 +57,17 @@ kotlin {
                 api(libs.touchlab.kermit)
                 api(libs.orbit.mvi.core)
                 api(libs.icerock.moko.mvvm.core)
+                api(libs.ballast.core)
```
Comment (seljabali) [inline shared/build.gradle.kts:60]: Remove:
```
api(libs.orbit.mvi.core)
```

IceRock too if we're not using it:
```
api(libs.icerock.moko.mvvm.core)
```

File: shared/src/commonMain/kotlin/co/touchlab/kampkit/metaweather/model/openweather/Coord.kt [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package co.touchlab.kampkit.metaweather.model.openweather
+
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class Coord(
```
Comment (seljabali) [inline shared/src/commonMain/kotlin/co/touchlab/kampkit/metaweather/model/openweather/Coord.kt:7]: Rename to `Coordinate` or `LatLong`

File: shared/src/commonMain/kotlin/co/touchlab/kampkit/metaweather/model/openweather/Sys.kt [hunk 7]
```diff
@@ -0,0 +1,22 @@
+package co.touchlab.kampkit.metaweather.model.openweather
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+
+@Serializable
+data class Sys(
```
Comment (seljabali) [inline shared/src/commonMain/kotlin/co/touchlab/kampkit/metaweather/model/openweather/Sys.kt:6]: `Sys`? Rename to something more descriptive.

File: shared/src/commonMain/kotlin/co/touchlab/kampkit/metaweather/model/openweather/WeatherForCity.kt [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+    val visibility: Long,
+
+    @SerialName("wind")
+    val wind: Wind,
+
+    @SerialName("clouds")
+    val clouds: Clouds,
+
+    @SerialName("dt")
+    val dt: Long,
```
Comment (seljabali) [inline shared/src/commonMain/kotlin/co/touchlab/kampkit/metaweather/model/openweather/WeatherForCity.kt:29]: Rename to something more descriptive.

File: shared/src/commonMain/kotlin/co/touchlab/kampkit/metaweather/model/openweather/WeatherForCity.kt [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    val timezone: Int,
+
+    @SerialName("id")
+    val id: Long,
+
+    @SerialName("name")
+    val cityName: String,
+
+    @SerialName("cod")
+    val cod: Int,
```
Comment (seljabali) [inline shared/src/commonMain/kotlin/co/touchlab/kampkit/metaweather/model/openweather/WeatherForCity.kt:44]: Rename to something more descriptive.

**Review by seljabali — CHANGES_REQUESTED** [review]
Looks good~!

File: shared/build.gradle.kts (hunk 5 above)
Comment (RamziJabali) [reply shared/build.gradle.kts:60]: I don't see them on github or locally

File: shared/src/commonMain/kotlin/co/touchlab/kampkit/metaweather/model/openweather/Coord.kt (hunk 6 above)
Comment (RamziJabali) [reply shared/src/commonMain/kotlin/co/touchlab/kampkit/metaweather/model/openweather/Coord.kt:7]: Sounds good!

File: app/build.gradle.kts [hunk 10]
```diff
@@ -27,8 +27,8 @@ android {
     }
 
     lint {
-        isWarningsAsErrors = false
-        isAbortOnError = false
+        isWarningsAsErrors = true
+        isAbortOnError = true
```
Comment (seljabali) [inline app/build.gradle.kts:31]: revert

File: app/build.gradle.kts [hunk 11]
```diff
@@ -41,17 +41,15 @@ android {
...
     }
 }
 
 dependencies {
     implementation(project(":shared"))
     implementation(libs.bundles.app.ui)
+    implementation(libs.multiplatformSettings.common)
+    implementation(libs.kotlinx.dateTime)
     coreLibraryDesugaring(libs.android.desugaring)
-    implementation(libs.koin.android)
```
Comment (seljabali) [inline app/build.gradle.kts:52]: removing koin?

File: docs/DEBUGGING_KOTLIN_IN_XCODE.md [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+### Kotlin Source in Xcode
+To take advantage of the plugin you will want to add references to your kotlin code in XCode. This will allow you to add breakpoints and edit kotlin without switching to Android Studio. You probably wont want to do your primary kotlin coding like this, but it's helpful when debugging.
+
+To add the Kotlin source:
+1. Right click in the project explorer
+
+![](Screenshots/AddFiles.png)
+
+2. In the finder opened, select the kotlin source you want included (commonMain and iosMain). Be sure to select "Create folder references for any added folders"
+
```
Comment (seljabali) [inline docs/DEBUGGING_KOTLIN_IN_XCODE.md:17]: Remove

File: docs/DETAILED_DEV_SETUP.md [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+You're KMP development environment is ready now. Your next step should be to go to the [APP_BUILD.md doc](APP_BUILD.md), which focuses on building this project, as well as running it on both Android and iOS.
+
+
+
+## Further Reading
+
+If you want a more in-depth guide to getting a KMP project up and running for those coming from an iOS background, check out [this walkthrough](https://www.infoq.com/articles/kotlin-multiplatform-ios-developers/) that we wrote (and are maintaining).
+
+
```
Comment (seljabali) [inline docs/DETAILED_DEV_SETUP.md:62]: Remove

File: gradle/libs.versions.toml [hunk 14]
```diff
@@ -6,44 +6,37 @@ compileSdk = "31"
...
-icerock-moko-mvvm-core = {module = "dev.icerock.moko:mvvm-core", version.ref = "moko"}
+androidx-test-junit = { module = "androidx.test.ext:junit-ktx", version = "1.1.3" }
 
+compose-compiler = { module = "androidx.compose.compiler:compiler", version.ref = "composeCompiler" }
 compose-ui = { module = "androidx.compose.ui:ui", version.ref = "compose" }
 compose-tooling = { module = "androidx.compose.ui:ui-tooling", version.ref = "compose" }
 compose-foundation = { module = "androidx.compose.foundation:foundation", version.ref = "compose" }
 compose-material = { module = "androidx.compose.material:material", version.ref = "compose" }
-compose-activity = { module = "androidx.activity:activity-compose", version = "1.3.1" }
-compose-navigation = { module = "androidx.navigation:navigation-compose", version.ref = "lifecycle"}
```
Comment (seljabali) [inline gradle/libs.versions.toml:39]: Revert

File: shared/src/commonMain/sqldelight/co/touchlab/kampkit/db/Table.sq [hunk 15]
```diff
@@ -14,11 +14,11 @@ selectByName:
...
+INSERT OR IGNORE INTO Breed(name)
+VALUES (?);
 
 deleteAll:
 DELETE FROM Breed;
 
 updateFavorite:
-UPDATE Breed SET favorite = ? WHERE id = ?;
\ No newline at end of file
+UPDATE Breed SET favorite = ? WHERE id = ?;
```
Comment (seljabali) [inline shared/src/commonMain/sqldelight/co/touchlab/kampkit/db/Table.sq:24]: Remove

**Comment by seljabali** [comment]
Remove the added readme pictures please

File: LICENSE.txt [hunk 16]
```diff
@@ -1,21 +0,0 @@
-MIT License
```
Comment (seljabali) [inline LICENSE.txt:1]: Revert deletion please.

**Comment by seljabali** [comment]
Revert files deleted in `docs/` as they're referenced in the README.

File: ios/KaMPKitiOS/AppDelegate.swift [hunk 17]
```diff
@@ -1,3 +1,11 @@
+//
+//  AppDelegate.swift
+//  KaMPKitiOS
+//
+//  Created by Kevin Schildhorn on 12/18/19.
+//  Copyright © 2019 Touchlab. All rights reserved.
+//
+
```
Comment (seljabali) [inline ios/KaMPKitiOS/AppDelegate.swift:8]: Delete

File: shared/src/commonMain/kotlin/co/touchlab/kampkit/metaweather/model/openweather/WeatherForCity.kt [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+    @SerialName("name")
+    val cityName: String,
+
+    @SerialName("cod")
+    val internalParameter: Int,
+
+
+
+
+
```
Comment (seljabali) [inline shared/src/commonMain/kotlin/co/touchlab/kampkit/metaweather/model/openweather/WeatherForCity.kt:50]: Remove extra white space

File: shared/src/commonMain/kotlin/co/touchlab/kampkit/metaweather/model/openweather/Wind.kt [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+
+@Serializable
+data class Wind(
+    @SerialName("speed")
+    val windSpeed: Double,
+
+    @SerialName("deg")
+    val deg: Double,
+
+    )
```
Comment (seljabali) [inline shared/src/commonMain/kotlin/co/touchlab/kampkit/metaweather/model/openweather/Wind.kt:14]: Format

File: docs/GENERAL_ARCHITECTURE.md [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+
+You may be thinking: Weren't the libraries injected? How does the dependency injection work? Well that is actually handled in the `TestUtil.kt` file in commonTest. In order to pass the platform specific libraries(SqlDelight requires a platform driver) into the BreedModel for testing we need to inject them. Before the tests in BreedModelTest we are calling `appStart` and passing in our libraries to be injected and starting Koin. Then after the tests we are stopping Koin by calling `appEnd`. Note that the actual `testDbConnection()` implementations are in the *SqlDelightTest.kt* files.
+
+#### Android
+On the android side we are using AndroidRunner to run the tests because we want to use android specifics in our tests. If you're not using android specific methods then you don't need to use AndroidRunner. The android tests are run can be easily run in Android Studio by right clicking on the shared folder, and selecting `Run 'All Tests'`.
+
+#### iOS
+iOS tests have their own gradle task allowing them to run with an iOS simulator. You can simply go to the terminal and run `./gradlew iosTest`.
```
Comment (seljabali) [inline docs/GENERAL_ARCHITECTURE.md:139]: Remove.

File: docs/TROUBLESHOOTING.md [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+**Q:** The Xcode project won't compile. On the `import shared` line in Swift, I'm getting a compilation error "no
+ such module: 'shared'".
+
+**A:** Try closing Xcode and deleting the `Pods/` folder located in the root directory of the iOS project. Then run the command `pod install` in that same iOS root directory (which is `/KaMPKit/ios/` to be specific). This command will generate a new `Pods` folder. Reopen the `.xcworkspace` file and try to build again.
+
+> Note: We're still not quite sure as to the cause of this error. Possible factors include differing versions of Cocoapods or >Xcode.
+
+## More to Come!
+
+[Let us know](../CONTACT_US.md) what issues you run into
```
Comment (seljabali) [inline docs/TROUBLESHOOTING.md]: Remove.

File: docs/WHAT_AND_WHY.md [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+* Non-UI: Picking a big, monolithic tech stack for mobile is risky. Shared UI doesn’t have a great history, but shared logic is the history of computers. Build something that is incremental and “plays nice” with the host system is much harder and will take longer, but is ultimately the successful strategy. There will be “shared UI” options for KMP. The good news is they’ll be optional.
+* Jetbrains: Jetbrains has built an amazing business on building the best IDE’s. They also make Kotlin the language. This is a unique combination. They are self-funded, as in there is no VC or public shareholder pressure to have 
+  immediate ROI. Jetbrains is here to stay, and they are committed to Kotlin. The tooling around KMP and Native is evolving, 
+  but it’s safe to assume Kotlin as a platform will have the best tooling in the industry.
+
+We, Touchlab, have a clear perspective on the future. That is, the future is very hard to predict. Kotlin as a platform 
+is a low-risk choice because of the reasons mentioned above. We prefer less risky choices because being “right” about 
+the future isn’t that important if you can reduce the cost of being “wrong”.
```
Comment (seljabali) [inline docs/WHAT_AND_WHY.md]: Remove.

File: docs/TROUBLESHOOTING.md (hunk 21 above)
Comment (RamziJabali) [reply docs/TROUBLESHOOTING.md]: I have already deleted this, not sure why it's showing as recently added

**Review by seljabali — APPROVED** [review]

**Pull request merged**